### PR TITLE
[APR] Tokens on pool page

### PR DESCRIPTION
**Pull request by Rawallon**
[Deployed](https://tools-monorepo-jbi2533os-ribeirojose.vercel.app/apr/pool/ethereum/0x6f0ed6f346007563d3266de350d174a831bde0ca0001000000000000000005db/round/74)

<details><summary>Balancer's</summary>

image

</details>

# Weighted Pools

image

# No round provided

image

# Before Pool's creation

image

When the pools is not weighted it doesnt show the "Volume" col

### CODE REVIEW

File: apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+}: {
+  poolStats: PoolStatsData;
+}) {
+  return (
+    <>
+      <Table color="blue" shade={"darkWithBorder"}>
+        <Table.HeaderRow>
+          <Table.HeaderCell>Name</Table.HeaderCell>
+          <Table.HeaderCell classNames="w-1/6 text-end p-4">
+            Volume
```
Comment (luizakp) [inline apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx:20]:
```suggestion
            Balance
```

**Comment by luizakp** [comment]
`#would` add `Token %` on the token table

Balancer example:

image

**Comment by luizakp** [comment]
It would be nice to add the token balance value in USD too, but this can be another task if needed

File: apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx (hunk 1 above)
Comment (Rawallon) [reply apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx:20]: I'm using balancer's nomenclature there

File: apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+              <Table.BodyCellLink
+                href={`https://etherscan.io/address/${token.address}`}
+                tdClassNames="w-6"
+                linkClassNames="justify-end w-full"
+              >
+                {parseFloat(token.balance).toLocaleString("en-US", {
+                  minimumFractionDigits: 0,
+                  maximumFractionDigits: 0,
+                })}
+              </Table.BodyCellLink>
```
Comment (luizakp) [inline apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx:73]: They are flipped
image

**Comment by luizakp** [comment]
`#must` add a title `Current Pool Composition`, otherwise a user might think that this composition during a selected round

**Comment by Rawallon** [comment]
> It would be nice to add the token balance value in USD too, but this can be another task if needed

There should already be a value for specific round, or do you mean for a single token?  We could also add the avg value for that token, but I think it would add a lot of aditional requests

**Comment by luizakp** [comment]
Handle when user select a round where the pool didn't exist

image

**Comment by luizakp** [comment]
> > It would be nice to add the token balance value in USD too, but this can be another task if needed
> 
> There should already be a value for specific round, or do you mean for a single token? We could also add the avg value for that token, but I think it would add a lot of aditional requests

hmmm in that case I suggest to break it like Balancer does. One column is tha balance, which is the amount of token. Then, value is that balance in USD, which is balance * current token price

File: apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx (hunk 1 above)
Comment (luizakp) [reply apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx:20]: can you provide an example? Maybe you are thinking about the "Volume" on the charts, but on the pool composition they use this nomenclature:

image

File: apps/balancer-tools/src/app/apr/pool/(components)/PoolTokens.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    poolData.perRound.map((pool) => {
+      pool.tokens.map(async (token) => {
+        // Is there a way to type this without recreating the interface?
+        // @ts-ignore 2345
+        const tprice = await getTokenPriceByRound(
+          Round.getRoundByNumber(roundId),
+          token.address,
+          parseInt(pool.network),
+        );
+        // @ts-ignore 2345
```
Comment (luizakp) [inline apps/balancer-tools/src/app/apr/pool/(components)/PoolTokens.tsx:30]: 👀

Comment (Rawallon) [reply apps/balancer-tools/src/app/apr/pool/(components)/PoolTokens.tsx:30]: There's no way around it is there?

File: apps/balancer-tools/src/app/apr/(utils)/calculatePoolStats.ts [hunk 4]
```diff
@@ -77,36 +81,59 @@ export async function calculatePoolStats({
...
+  ]);
+
+  let tokens;
+  try {
+    tokens = pool.tokens.map((token, idx) => ({
+      logo: token.logo,
+      address: token.address,
+      symbol: token.symbol,
+      weight: token.weight,
+      balance: tokenBalance[idx].balance,
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/(utils)/calculatePoolStats.ts:126]:
```suggestion
      balance: tokenBalance?.[idx]?.balance,
```

File: apps/balancer-tools/src/app/apr/(utils)/getBALPriceByRound.ts [hunk 5]
```diff
@@ -1,16 +1,30 @@
...
+    1,
+  );
+};
+
+export const getTokenPriceByRound = async (
+  round: Round,
+  tokenAddress: string,
+  tokenNetwork: number,
+) => {
+  const token = `${networkFor(tokenNetwork)}:${tokenAddress}`;
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/(utils)/getBALPriceByRound.ts:20]: this first part must be lowercase

File: apps/balancer-tools/src/app/apr/pool/(components)/PoolTokens.tsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      const tokenPrice = await getTokenPriceByRound(
+        Round.getRoundByNumber(roundId),
+        token.address,
+        parseInt(poolData.perRound[0].network),
+      );
+      token.price = tokenPrice;
+      token.percentageValue =
+        ((tokenPrice * parseFloat(token.balance)) / totalBalance) * 100;
+    });
+  }
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/pool/(components)/PoolTokens.tsx:38]: any reason not to return this with the API response, or maybe throw this to a different endpoint?

File: apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+export default function PoolTokensTable({
+  poolTokensStats,
+}: {
+  poolTokensStats: PoolTokens[];
+}) {
+  const tokenBalanceUSD = (balance: string) =>
+    parseFloat(balance).toLocaleString("en-US", {
+      minimumFractionDigits: 0,
+      maximumFractionDigits: 0,
+    });
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx:18]: IMO we should return early if there's nothing in `poolTokensStats`

File: apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            <Table.HeaderCell classNames="w-1/6 text-end p-4">
+              Token %
+            </Table.HeaderCell>
+          )}
+        </Table.HeaderRow>
+        <Table.Body>
+          {poolTokensStats.map((token) => (
+            <Table.BodyRow classNames="hover:bg-blue4 hover:cursor-pointer duration-500">
+              <Table.BodyCellLink
+                href={`https://etherscan.io/address/${token.address}`}
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx:49]: tokens won't all be in etherscan -- remember this is multichain. I think we have a canonical list of chain scanners somewhere

File: apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+                </Table.BodyCellLink>
+              )}
+              {token.price && poolTokensStats[0].balance && (
+                <Table.BodyCellLink
+                  href={`https://etherscan.io/address/${token.address}`}
+                  tdClassNames="w-6"
+                  linkClassNames="justify-end w-full"
+                >
+                  {`$${(
+                    token.price! * parseFloat(token.balance)
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx:94]: why use the non null assertion (!) here?

**Review by ribeirojose — APPROVED** [review]
lgtm only a few sharp edges left. Is this already mobile-friendly?

File: apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx (hunk 9 above)
Comment (Rawallon) [reply apps/balancer-tools/src/app/apr/pool/(components)/PoolTokensTable.tsx:94]: There's a `if token.price && poolTokensStats[0].balance`, if that code gets reached should mean that it isn't undefined

**Pull request merged**